src/directory/s3_directory.rs

```diff
@@ -0,0 +1,248 @@
+
+#[derive(Clone)]
+struct InnerDirectory(Arc<RwLock<HashMap<PathBuf, Arc<Vec<u8>>>>>);
+
+impl InnerDirectory {
+    fn new() -> InnerDirectory {
+        InnerDirectory(Arc::new(RwLock::new(HashMap::new())))
+    }
+}
+
+/// Directory storing data in files, read via mmap.
+///
+/// The Mmap object are cached to limit the
+/// system calls.
+#[derive(Clone)]
+pub struct S3Directory {
+    root_path: PathBuf,
+    bucket: String,
+    region: Region,
+    fs: InnerDirectory,
+}
+
+impl fmt::Debug for S3Directory {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "S3Directory({:?})", self.root_path)
+    }
+}
+
+impl S3Directory {
+    /// Opens a S3Directory in a bucket.
+    ///
+    /// Returns an error if the `bucket` does not
+    /// exist or if it is not a directory.
+    pub fn open(
+        region: String,
+        bucket: String,
+        directory_path: &Path,
+    ) -> Result<S3Directory, OpenDirectoryError> {
+        // TODO: should I use a different error type? probably
+
+        let region = Region::from_str(&region).map_err(|_| {
+            OpenDirectoryError::DoesNotExist(PathBuf::from("/bad/region"))
+        })?;
+
+        // TODO: handle missing creds
+        let client = default_tls_client().map_err(|_| {
+            OpenDirectoryError::DoesNotExist(PathBuf::from("/bad/tls/client"))
+        })?;
+
+        let provider = DefaultCredentialsProvider::new().map_err(|_| {
+            OpenDirectoryError::DoesNotExist(PathBuf::from("/bad/creds"))
+        })?;
+
+        // TODO: do I want to save this off?
+        let s3 = S3Client::new(client, provider, region.clone());
+
+        // does bucket exist?
+        s3.head_bucket(&HeadBucketRequest { bucket: bucket.clone() })
+            .map_err(|_| {
+                OpenDirectoryError::DoesNotExist(PathBuf::from("/no/bucket"))
+            })?;
+
+        // TODO: how to store the client?
+        Ok(S3Directory {
+            bucket,
+            region: region,
+            root_path: PathBuf::from(directory_path),
+            fs: InnerDirectory::new(),
+        })
+
+    }
+
+    fn get_client(&self) -> Box<S3> {
+        let client = default_tls_client()
+            .map_err(|_| {
+                OpenReadError::FileDoesNotExist(PathBuf::from("/bad/tls/client"))
+            })
+            .unwrap();
+
+        let provider = DefaultCredentialsProvider::new()
+            .map_err(|_| {
+                OpenReadError::FileDoesNotExist(PathBuf::from("/bad/creds"))
+            })
+            .unwrap();
+
+        Box::new(S3Client::new(client, provider, self.region.clone()))
+    }
+
+    /// Joins a relative_path to the directory `root_path`
+    /// to create a proper complete `filepath`.
+    fn resolve_path(&self, relative_path: &Path) -> PathBuf {
+        self.root_path.join(relative_path)
+    }
+}
+
+impl Directory for S3Directory {
+    fn open_read(&self, path: &Path) -> result::Result<ReadOnlySource, OpenReadError> {
```

Comment:
This method has been implemented in a very brute force procedural way. My goals are to make sure I understand all of the primitives I'm going to be using. I'll then look at extracting methods to better leverage rust's option/result chaining methods.

If you can help invest in my understanding of rust here, I feel that it will carry out through the rest of the code.